File: extract_events_start.py
import pandas as pd
from typing import Dict, Any
from str2num_b1b2b3 import str2num_b1b2b3
from eval_hit_events import eval_hit_events
from extract_events import extract_events
from eval_point_and_kiss_control import eval_point_and_kiss_control
from create_varname import create_varname


def extract_events_start(SA: dict, param: dict):
    """
    Extract all ball-ball hit events and ball-cushion hit events
    
    Args:
        SA: Dictionary containing shot data with DataFrames
        player: Dictionary containing player information
    """
    print(f"start (extract_events_start)")

    # Check if B1B2B3 column exists
    if 'B1B2B3' not in SA['Table'].columns:
        print("B1B2B3 is not identified yet (extract_events_start)")
        return

    err = 0
    err_shots = []

    shot_length = len(SA['Table'])
    for si, row in SA['Table'].iterrows():
        if row['Interpreted'] == 0:
            print(f"Shot {si+1}/{shot_length}")
            if len(row['B1B2B3']) == 3:
                # try:
                # Convert B1B2B3 string to indices
                b1b2b3, b1i, b2i, b3i = str2num_b1b2b3(row['B1B2B3'])
                
                # Extract all events
                hit, _ = extract_events(si, SA, param)
                
                # Evaluate hit events
                hit = eval_hit_events(si, hit, b1b2b3, SA, param)
                
                # Evaluate point and kiss control
                hit = eval_point_and_kiss_control(si, hit, SA, param)
                
                # Create variables in SA Table
                SA = create_varname(si, hit, SA, param)
                
                # Store hit data
                if 'Shot' not in SA:
                    SA['Shot'] = {}
                if si not in SA['Shot']:
                    SA['Shot'][si] = {}
                SA['Shot'][si]['hit'] = hit
                
                # Update flags
                SA['Table'].at[si, 'Interpreted'] = 1
                SA['Table'].at[si, 'ErrorID'] = None
                SA['Table'].at[si, 'ErrorText'] = ''
                SA['Table'].at[si, 'Selected'] = False
                
                # Clear route data
                if 'Route' in SA and si in SA['Route']:
                    SA['Route'].drop(SA['Route'][SA['Route']['ShotID'] == row['ShotID']].index, inplace=True)
                    
                # except Exception as e:
                #     SA['Table'].at[si, 'ErrorID'] = 100
                #     SA['Table'].at[si, 'ErrorText'] = 'Check diagram, correct or delete.'
                #     SA['Table'].at[si, 'Selected'] = True
                    
                #     print("Some error occurred, probably ball routes are not continuous. Check diagram, correct or delete")
                #     print(f"Error: {str(e)}")
                #     err += 1
                #     err_shots.append(si)
            else:
                print(f"B1B2B3 has not 3 letters, skipping Shot {si+1}")

    print("These Shots are not interpreted:")
    for shot in err_shots:
        print(shot)
    
    print("done (extract_events_start)")

File: Extract_Events_start.m
function Extract_Events_start(~,~)

% Extract all ball-ball hit events and ball-Cushion hit events

global SA player

if isempty(find(strcmp(SA.Table.Properties.VariableNames,'B1B2B3'),1))
    disp(['B1B2B3 is not identified yet (',mfilename,')'])
    return
end

err = 0;
err_shots = [];

shotlength = length(SA.Shot);
for si = 1:shotlength
    if SA.Table.Interpreted{si} == 0 
        disp(['Shot ', num2str(si),'/',num2str(shotlength)])
        if length(SA.Table.B1B2B3{si}) == 3
            % when there is cushion contact in next time step
            %       table_fig = findobj('Tag','Table_figure');
            %       delete(table_fig)
            %
            %       SAD.selectedrows = si;
            %       htable_shotlist = findobj('Tag', 'uitable_shotlist');
            %       set(htable_shotlist, 'Userdata', SAD);
            %       plot_selection
            
            try
                [b1b2b3, b1i, b2i, b3i] = str2num_B1B2B3(SA.Table.B1B2B3{si});
                
                % extract all events
                [hit, ~] = Extract_Events(si);
                
                % collect the events
                hit = eval_hit_events(hit,si,b1b2b3);

                % This is now done in create_varname.m
%                 % replace b1b2b3 codes
%                 hit = replace_hitcode_b1b2b3(hit, b1b2b3, b1i, b2i, b3i);

                % evaluate hit accuracy and kiss control
                hit = eval_Point_and_Kiss_Control(si, hit);
                
                % create the SA.Table
                SA = create_varname(SA, hit, si) ;
                
                % copy the hit data to SA
                SA.Shot(si).hit = hit;
                
                % Set intrepeted flag
                SA.Table.Interpreted{si} = 1;
                
                % reset flags and delete Route data
                SA.Table.ErrorID{si} = [];
                SA.Table.ErrorText{si} = [];
                SA.Table.Selected{si} = false;
                SA.Shot(si).Route = [];
                
                
            catch
                SA.Table.ErrorID{si} = 100;
                SA.Table.ErrorText{si} = 'Check diagram, correct or delete.';
                SA.Table.Selected{si} = true;
                
                disp('Some error occured, probably ball routes are not continous. Check diagram, correct or delete')
                err = err + 1;
                err_shots(err) = si;
            end
        else
            disp(['B1B2B3 has not 3 letters, skipping Shot ',num2str(si)])
        end
    end
    
end

update_ShotList

disp('These Shots are not interpreted:')
for i=1:err
    disp(num2str(err_shots(i)))
end
% update GUI
update_ShotList
player.uptodate = 0;
PlayerFunction('plotcurrent',[])

disp(['done (',mfilename,')'])
